Just keep one condition for each program object

anfredette opened the pull request:
All of the code assumes that only the last condition counts.  However, that's not how conditions work.  This change modifies our behavior so that we only maintain the one (most recent) condition for each program object, which is a step toward the resolution of issue #430.

Part of the solution is to create a common updateCondition() function for the bpfd-operator that is used for all program types which removes some code duplication and is a small step toward the resolution of issue #709

This also removes code duplication in xdp-program_test.go, which is a small step toward the resolution of issue #710.

astoycos commented on bpfd-operator/controllers/bpfd-agent/common.go:
```diff
@@ -164,25 +164,50 @@ func (r *ReconcilerCommon) removeFinalizer(ctx context.Context, o client.Object,
+					meta.RemoveStatusCondition(&prog.Status.Conditions, prog.Status.Conditions[0].Type)
+				}
+			}
+		case numConditions == 0:
+			{
+				// Fall through and add the new condition.
+			}
```
nit remove this, just use `default:` and print 	`r.Logger.V(1).Info("Updating bpfProgram condition", "bpfProgram", prog.Name, "condition", cond.Condition().Type)`

astoycos commented on bpfd-operator/controllers/bpfd-agent/xdp-program_test.go:
```diff
@@ -43,184 +44,34 @@ import (
 	)
+
+	var fakeInts []string
+	if multiInterface {
```
This if statement always does the same thing :/

anfredette replied:
oops. fixed.

astoycos commented on bpfd-operator/controllers/bpfd-agent/xdp-program_test.go:
```diff
@@ -424,43 +198,153 @@ func TestXdpProgramControllerCreateMultiIntf(t *testing.T) {
+	}
+}
+
+func TestXdpProgramControllerCreate(t *testing.T) {
```
Don't we need to make these same changes for the tc program unit tests as well?

anfredette replied:
We could, but I wasn't changing the tc cases, so I didn't do it.

astoycos commented on bpfd-operator/controllers/bpfd-operator/common.go:
```diff
@@ -78,14 +80,14 @@ func reconcileBpfProgram(ctx context.Context, rec ProgramReconciler, prog client
 	nodes := &corev1.NodeList{}
 	if err := r.List(ctx, nodes, &client.ListOptions{}); err != nil {
-		r.Logger.Error(err, "failed getting nodes for full reconcile")
+		r.Logger.V(1).Error(err, "failed getting nodes for full reconcile")
```
This is the only place where we see this error, I don't think you should bump these to V(1)

anfredette replied:
fixed

astoycos commented on bpfd-operator/controllers/bpfd-operator/common.go:
```diff
@@ -188,3 +190,50 @@ func statusChangedPredicate() predicate.Funcs {
 	}
 }
+
+func (r *ReconcilerCommon) updateCondition(ctx context.Context, obj client.Object, conditions *[]metav1.Condition, cond bpfdiov1alpha1.ProgramConditionType, message string) (ctrl.Result, error) {
```
Can you move this and the one in bpfd-agent/common.go into `/bpfd-operator/internal/k8s.go` ?

astoycos commented on bpfd-operator/controllers/bpfd-operator/common.go:
```diff
@@ -188,3 +190,50 @@ func statusChangedPredicate() predicate.Funcs {
+	meta.SetStatusCondition(conditions, cond.Condition(message))
+
+	if err := r.Status().Update(ctx, obj); err != nil {
+		r.Logger.V(1).Info("failed to set KprobeProgram object status...requeuing")
```
```suggestion
		r.Logger.V(1).Info("failed to set *Program object status...requeuing")
```

anfredette replied:
good catch

astoycos commented on bpfd-operator/controllers/bpfd-operator/kprobe-program_test.go:
```diff
@@ -137,6 +140,21 @@ func TestKprobeProgramReconcile(t *testing.T) {
+			r.Logger.V(1).Info("failed to set KprobeProgram object status")
+		}
+		// Make sure we have 2 conditions
+		require.Equal(t, 2, len(Kprobe.Status.Conditions))
```
Wait shouldn's we only have one condition?

anfredette replied:
This is a forced error test case where I explicitly added 2 conditions. This should never happen with the current code, but I wanted to handle it in case someone messes it up in the future.

anfredette replied:
Also, since all the program types run the same code, I just randomly picked one in which to test this error path.

astoycos reviewed (changes requested):
just some small stuff!! Let's condense those functions that the agent + operator are using to one central location if possible.

astoycos commented:
Shoot I forgot about https://github.com/bpfd-dev/bpfd/blob/main/bpfd-operator/pkg/helpers/helpers.go#L308

can you update all those helpers to not play the "look for last position" dance?

anfredette commented:
> Shoot I forgot about https://github.com/bpfd-dev/bpfd/blob/main/bpfd-operator/pkg/helpers/helpers.go#L308
> 
> can you update all those helpers to not play the "look for last position" dance?

Thanks for pointing that out.  I updated the code and moved the repeated condition-checking logic into an isProgLoaded() function.

astoycos commented on bpfd-operator/pkg/helpers/helpers.go:
```diff
@@ -305,6 +305,31 @@ func GetMaps(c *bpfdclientset.Clientset, ProgramName string, mapNames []string)
+		return false, "None"
+	}
+
+	if conLen > 1 {
```
We've decided in our core design that we should only ever have one condition. 

I don't think we should be "assuming" this may not hold true in a public library that we've designed for people to use. 

i.e this function should return the single condition `conditions[0]` and return (false, error)  or (false, "None") otherwise

anfredette replied:
That's what it's doing -- I'm just logging a message if the condition exists.  However, I'm okay with removing the check and log here in the public library.

astoycos commented on bpfd-operator/controllers/bpfd-operator/common.go:
```diff
@@ -188,3 +190,46 @@ func statusChangedPredicate() predicate.Funcs {
+				// this case.  However, if we do, delete the existing conditions
+				// and add the new one below.
+				r.Logger.Info("more than one BpfProgramCondition", "numConditions", numConditions)
+				for _, c := range *conditions {
```
^^ is just https://github.com/kubernetes/apimachinery/blob/v0.28.2/pkg/api/meta/conditions.go#L30 with a single line changed to make the operation idempotent rather than additive

anfredette replied:
First of all, this should never happen, so we shouldn't execute this code and there shouldn't be any logs.  However, in my experience, I've seen many bugs in code that had comments saying "this should never happen".  If you ignore the possiblities, you wind up having bugs that are very hard to find.  And this issue isn't with how it works now.  The issue would happen a year or two from now when others are working on the project and someone decides it would be good to report a condition for something new and it messes up how the core logic works.  So, this is defensive programming.  If it happens, we log a message and clean things up.

anfredette replied:
I think we can remove the debug logs that say we're removing the condtions since we already have the debug log at the beginning that reports the existing and new conditions.  I think it's good to at least have this log so we can understand how the state of the objects is changing if something goes wrong.

astoycos replied:
OK i'm alright with it I just didn't like the extra loop :)

anfredette commented on bpfd-operator/pkg/helpers/helpers.go:
```diff
@@ -437,10 +426,10 @@ func WaitForBpfProgConfLoad(c *bpfdclientset.Clientset, progName string, timeout
-	// the only case currently used is Xdp, so it's not needed now.  If we need
-	// to support uprobes in the future, we could replace ProgramType with
-	// something else, like the string representation of the program type.
+	// TODO: case Uprobe: not covered.  Since Uprobe has the same ProgramType as
```
Good idea.  Now or later?

astoycos reviewed (approved):
LGTM 

Barring response to Billy's review comments + his LGTM

anfredette commented:
I changed the logs and made a few other changes too.

anfredette commented:
I removed the hold.  It's waiting on @Billy99 to ack the changes.

Billy99 reviewed (approved):
LGTM
